### ISSUE TITLE
extract region features

### ISSUE THREAD

**Issue by mactavish91** [microsoft/RegionCLIP#30]
Hello，when running **extract_region_features.sh**, there is a strange bug in **extract_region_features.py**:
     at line 113 _extract_region_features.py predictions = model.roi_heads.box_predictor(att_feats)_ ,The model generated 100 prediction results. However, when the code in line 114 _pred_instances, keep_indices = model.roi_heads.box_predictor.inference(predictions, proposals) # apply per-class NMS_  is executed, the total number of output results changes from 100 to 300. Instead of achieving the effect of nms, many boxes that are completely coincident are generated.

Partially generated box results：
![image](https://user-images.githubusercontent.com/58135925/191564808-ef04a246-74fd-4884-9663-aa8c4ff724c0.png)

**Comment by YiwuZhong** [comment]
@mactavish91 I haven't met the problem you had here. To better understand the bug you mentioned, do you mind sharing the script/command you run?

**Comment by mactavish91** [comment]
>

**Comment by mactavish91** [comment]
> @mactavish91 I haven't met the problem you had here. To better understand the bug you mentioned, do you mind sharing the script/command you run?

@YiwuZhong I just follow the script and didn't change anything, the code runs as follows:
![image](https://user-images.githubusercontent.com/58135925/191645496-71458d0a-a222-4c18-8c4c-1cf229d826a2.png)

![image](https://user-images.githubusercontent.com/58135925/191645438-36e17528-c164-4c3c-bb69-77cbf30f2629.png)

The variable **prediction** contains 100 proposals, but the variable **pred_instances** returned by function **inference** generated 300 instances, with a lot of duplicates.

**Comment by mactavish91** [comment]
> @mactavish91 I haven't met the problem you had here. To better understand the bug you mentioned, do you mind sharing the script/command you run?

@YiwuZhong 
![image](https://user-images.githubusercontent.com/58135925/191670057-4605c385-c2a1-4d45-b606-0fdfbc7d843a.png)
Visual results from running code **extract_region_features.py**，it seems that NMS is not running correctly. Another problem is that the pretrain model (without fine-tuning on coco and lvis) can hardly recognize person, and the original clip model also has this problem. Do you know why this happens? Thank you very much.

**Comment by mactavish91** [comment]
@YiwuZhong At this time, I use region-clip finetune-livs instead of pretraining model, the result is much more accurate, but the NMS still does not work properly, and all person targets cannot be identified.
![image](https://user-images.githubusercontent.com/58135925/191735705-c3d94e02-5cd7-498e-a162-e8853d74db2a.png)
raw image:
![308](https://user-images.githubusercontent.com/58135925/191737543-e2d11d6b-c930-4808-b12a-59b2894f1f62.jpg)

**Comment by YiwuZhong** [comment]
@mactavish91 Your reported problem didn't show up when I ran the **default** ``extract_region_features.sh`` on your image. **predictions** has 1000 boxes, **pred_instances** has 100 boxes, and NMS works correctly.

After a closer look at your config, I realized that you changed the config and that's why you met the error. First, you set `OFFLINE_RPN_POST_NMS_TOPK_TEST 100` so that RPN only outputs 100 boxes (**predictions**). Second, you removed `TEST.DETECTIONS_PER_IMAGE 100` in ``extract_region_features.sh``. That's why you got **pred_instances** to be 300 boxes (300 is the default value for LVIS configs).

Your visualization only shows around 10 boxes and the person boxes might appear when you use the correct config. It's possible that both CLIP and RegionCLIP are not sensitive to `person` during zero-shot inference. Both models were trained by sentence description where `person` might not appear frequently. Actually, `person` is a vague concept, including `boy`, `girl`, `woman`, `man` and etc. But after fine-tuning on human annotation (many instances of `person`), the model can correctly recognize it.

**Comment by mactavish91** [comment]
@YiwuZhong Thank you very much for your kindness help, you are the light of the dark night, illuminating the direction of my moving forward. However, there is still one thing that bothers me: when training the detector, the model only has the label information of the base class, and the parameters of the image encoder will be updated during this process. Will this process destroy the representation of the novel class and thus reduce the classification effect of the new class?
